File: tools/create-model/create-model.py
#!/usr/bin/env python3
import numpy as np
from gguf import GGUFWriter, GGUFValueType
import sys
import os

def create_gguf_file(filename: str):
    """
    Creates a GGUF file with the same structure as the provided C++ code.
    """
    absolute_path = os.path.join(os.getcwd(), filename)
    arch = "llama"
    writer = GGUFWriter(filename, arch)

    print(f"Creating GGUF file '{filename}' with architecture '{arch}'")

    # --- Add Required Hyperparameters ---
    VOCAB_SIZE = 3
    BLOCK_COUNT = 1
    CONTEXT_LENGTH = 512
    EMBEDDING_LENGTH = 128
    FEED_FORWARD_LENGTH = 512
    HEAD_COUNT = 8
    HEAD_COUNT_KV = 1
    LAYER_NORM_RMS_EPS = 1e-5

    writer.add_string("general.name", "test-gguf")
    writer.add_uint32("llama.vocab_size", VOCAB_SIZE)
    writer.add_uint32("llama.block_count", BLOCK_COUNT)
    writer.add_uint32("llama.context_length", CONTEXT_LENGTH)
    writer.add_uint32("llama.embedding_length", EMBEDDING_LENGTH)
    writer.add_uint32("llama.feed_forward_length", FEED_FORWARD_LENGTH)
    writer.add_uint32("llama.attention.head_count", HEAD_COUNT)
    writer.add_uint32("llama.attention.head_count_kv", HEAD_COUNT_KV)
    writer.add_float32("llama.attention.layer_norm_rms_epsilon", LAYER_NORM_RMS_EPS)

    # --- Define Tokenizer (using specific methods and plain Python lists) ---
    tokens = ["<unk>", "<s>", "</s>"]
    scores = [0.0, 0.0, 0.0]
    toktypes = [2, 3, 3]

    writer.add_tokenizer_model("llama")
    writer.add_token_list(tokens)
    writer.add_token_scores(scores)
    writer.add_token_types(toktypes)

    # --- Define and Add Tensors with Random Data ---
    print("Defining and adding model tensors with random data...")
    def add_random_tensor(name: str, shape: tuple[int, ...]):
        tensor = np.random.uniform(-0.1, 0.1, size=shape).astype(np.float32)
        writer.add_tensor(name, tensor)

    add_random_tensor("token_embd.weight", (VOCAB_SIZE, EMBEDDING_LENGTH))
    add_random_tensor("output.weight", (VOCAB_SIZE, EMBEDDING_LENGTH))
    add_random_tensor("output_norm.weight", (EMBEDDING_LENGTH,))
    add_random_tensor("blk.0.attn_norm.weight", (EMBEDDING_LENGTH,))
    add_random_tensor("blk.0.attn_q.weight", (EMBEDDING_LENGTH, EMBEDDING_LENGTH))
    add_random_tensor("blk.0.attn_k.weight", (EMBEDDING_LENGTH // HEAD_COUNT * HEAD_COUNT_KV, EMBEDDING_LENGTH ))
    add_random_tensor("blk.0.attn_v.weight", (EMBEDDING_LENGTH // HEAD_COUNT * HEAD_COUNT_KV, EMBEDDING_LENGTH ))
    add_random_tensor("blk.0.attn_output.weight", (EMBEDDING_LENGTH, EMBEDDING_LENGTH))
    add_random_tensor("blk.0.ffn_norm.weight", (EMBEDDING_LENGTH,))
    add_random_tensor("blk.0.ffn_gate.weight", (FEED_FORWARD_LENGTH, EMBEDDING_LENGTH, ))
    add_random_tensor("blk.0.ffn_down.weight", (EMBEDDING_LENGTH, FEED_FORWARD_LENGTH))
    add_random_tensor("blk.0.ffn_up.weight", (FEED_FORWARD_LENGTH, EMBEDDING_LENGTH))
    
    # --- Write GGUF File (Multi-step process based on the example) ---
    print(f"\nWriting file step-by-step to: {absolute_path}")
    writer.write_header_to_file()
    writer.write_kv_data_to_file()
    writer.write_tensors_to_file()
    writer.close() 
    
    print(f"✅ GGUF test model written successfully.")
    
    # Final check to confirm file existence
    if os.path.exists(absolute_path):
        print(f"Confirmed: File '{filename}' now exists in the current directory.")
    else:
        print(f"Warning: File '{filename}' still not found after write operations.")


if __name__ == "__main__":
    create_gguf_file("test-gguf.gguf")

File: tools/create-model/create-model.cpp
#include "ggml.h"
#include "gguf.h"
#include <stdio.h>
#include <stdlib.h>
#include <cstring>

// Helper function to fill a tensor with random data
static void fill_random_f32(struct ggml_tensor *tensor) {
    if (tensor == NULL) return;
    size_t n = ggml_nelements(tensor);
    float *data = (float *) tensor->data;
    for (size_t i = 0; i < n; i++) {
        data[i] = ((float) rand() / RAND_MAX - 0.5f) * 0.2f;
    }
}

int main() {
    // --- Initialize GGUF Context ---
    struct ggml_init_params params = { 128 * 1024 * 1024, NULL, false };
    struct ggml_context *ctx_data = ggml_init(params);
    struct gguf_context *ctx_gguf = gguf_init_empty();

    // --- Set Metadata ---
    gguf_set_val_str(ctx_gguf, "general.architecture", "llama");
    gguf_set_val_str(ctx_gguf, "general.name", "test-gguf");

    // --- Add Required Hyperparameters ---
    gguf_set_val_u32(ctx_gguf, "llama.vocab_size", 3);                // 词汇表大小
    gguf_set_val_u32(ctx_gguf, "llama.block_count", 1);              // Transformer block 数量
    gguf_set_val_u32(ctx_gguf, "llama.context_length", 512);         // 上下文长度
    gguf_set_val_u32(ctx_gguf, "llama.embedding_length", 128);       // 嵌入维度
    gguf_set_val_u32(ctx_gguf, "llama.feed_forward_length", 512);    // 前馈网络维度
    gguf_set_val_u32(ctx_gguf, "llama.attention.head_count", 8);     // 注意力头数量
    gguf_set_val_u32(ctx_gguf, "llama.attention.head_count_kv", 1);  // KV 注意力头数量
    gguf_set_val_f32(ctx_gguf, "llama.attention.layer_norm_rms_epsilon", 1e-5f); // LayerNorm RMS Epsilon

    // --- Define Tokenizer ---
    const char *tokens[] = { "<unk>", "<s>", "</s>" };
    float scores[] = { 0.0f, 0.0f, 0.0f };
    int32_t toktypes[] = { 2, 3, 3 };

    gguf_set_val_str(ctx_gguf, "tokenizer.ggml.model", "llama");
    gguf_set_arr_str(ctx_gguf, "tokenizer.ggml.tokens", tokens, 3);
    gguf_set_arr_data(ctx_gguf, "tokenizer.ggml.scores", GGUF_TYPE_FLOAT32, scores, 3);
    gguf_set_arr_data(ctx_gguf, "tokenizer.ggml.token_type", GGUF_TYPE_INT32, toktypes, 3);

    // --- Add token_embd.weight Tensor ---
    int vocab_size = 3; // 词汇表大小
    int embedding_dim = 128; // 嵌入维度

    // --- Helper for adding tensors ---
    auto add_tensor = [&](const char *name, struct ggml_tensor *tensor) {
        fill_random_f32(tensor);
        ggml_set_name(tensor, name);
        gguf_add_tensor(ctx_gguf, tensor);
    };

    printf("Defining model tensors...\n");
    add_tensor("token_embd.weight", ggml_new_tensor_2d(ctx_data, GGML_TYPE_F32, embedding_dim, vocab_size));
    add_tensor("output.weight", ggml_new_tensor_2d(ctx_data, GGML_TYPE_F32, embedding_dim, vocab_size));
    add_tensor("output_norm.weight", ggml_new_tensor_1d(ctx_data, GGML_TYPE_F32, embedding_dim));
    add_tensor("blk.0.attn_norm.weight", ggml_new_tensor_1d(ctx_data, GGML_TYPE_F32, embedding_dim));
    add_tensor("blk.0.attn_q.weight",    ggml_new_tensor_2d(ctx_data, GGML_TYPE_F32, embedding_dim, embedding_dim));
    add_tensor("blk.0.attn_k.weight",    ggml_new_tensor_2d(ctx_data, GGML_TYPE_F32, embedding_dim, embedding_dim/8));
    add_tensor("blk.0.attn_v.weight",    ggml_new_tensor_2d(ctx_data, GGML_TYPE_F32, embedding_dim, embedding_dim/8));
    add_tensor("blk.0.attn_output.weight",    ggml_new_tensor_2d(ctx_data, GGML_TYPE_F32, embedding_dim, embedding_dim));
    add_tensor("blk.0.ffn_norm.weight", ggml_new_tensor_1d(ctx_data, GGML_TYPE_F32, embedding_dim));
    add_tensor("blk.0.ffn_gate.weight", ggml_new_tensor_2d(ctx_data, GGML_TYPE_F32, embedding_dim, embedding_dim*4));
    add_tensor("blk.0.ffn_down.weight", ggml_new_tensor_2d(ctx_data, GGML_TYPE_F32, embedding_dim*4, embedding_dim));
    add_tensor("blk.0.ffn_up.weight",   ggml_new_tensor_2d(ctx_data, GGML_TYPE_F32, embedding_dim, embedding_dim*4));

    // --- Write GGUF File ---
    const char* filename = "test-gguf.gguf";
    printf("Writing to file '%s'...\n", filename);
    gguf_write_to_file(ctx_gguf, filename, false);

    printf("✅ GGUF test model written successfully.\n");
    ggml_free(ctx_data);
    gguf_free(ctx_gguf);

    return 0;
}